File: test/Datadog.Trace.Tests.Properties/IntegrationVersionRangeTests.fs
module Datadog.Trace.Tests.Properties.Tests.ClrProfiler.Managed.IntegrationVersionRangeTests

open System
open FsCheck.Xunit
open Datadog.Trace.ClrProfiler

// adapted from /test/Datadog.Trace.ClrProfiler.Managed.Tests/IntegrationVersionRangeTests.cs

type Version =
    { Major: uint16
      Minor: uint16
      Patch: uint16 }

[<Property>]
let ``Minimum version two sets resets defaults for non specified parts`` (oldVer: Version) (newVer: Version) =
    let range = new IntegrationVersionRange()
    range.MinimumVersion <- $"{oldVer.Major}.{oldVer.Minor}.{oldVer.Patch}"
    range.MinimumVersion <- $"{newVer.Major}"
    newVer.Major = range.MinimumMajor &&
    0us = range.MinimumMinor &&
    0us = range.MinimumPatch

[<Property>]
let ``Parses minimum major`` (ver: Version) =
    let range = new IntegrationVersionRange()
    range.MinimumVersion <- $"{ver.Major}.{ver.Minor}"
    ver.Major = range.MinimumMajor

[<Property>]
let ``Parses minimum major and minor`` (ver: Version) =
    let range = new IntegrationVersionRange()
    range.MinimumVersion <- $"{ver.Major}.{ver.Minor}"
    ver.Major = range.MinimumMajor &&
    ver.Minor = range.MinimumMinor

[<Property>]
let ``Parses minimum major and minor and patch`` (ver: Version) =
    let range = new IntegrationVersionRange()
    range.MinimumVersion <- $"{ver.Major}.{ver.Minor}.{ver.Patch}"
    ver.Major = range.MinimumMajor &&
    ver.Minor = range.MinimumMinor &&
    ver.Patch = range.MinimumPatch

[<Property>]
let ``Maximum version two sets resets defaults for non specified parts`` (oldVer: Version) (newVer: Version) =
    let range = new IntegrationVersionRange()
    range.MaximumVersion <- $"{oldVer.Major}.{oldVer.Minor}.{oldVer.Patch}"
    range.MaximumVersion <- $"{newVer.Major}"
    newVer.Major = range.MaximumMajor &&
    UInt16.MaxValue = range.MaximumMinor &&
    UInt16.MaxValue = range.MaximumPatch

[<Property>]
let ``Parses maximum major`` (ver: Version) =
    let range = new IntegrationVersionRange()
    range.MaximumVersion <- $"{ver.Major}"
    ver.Major = range.MaximumMajor

[<Property>]
let ``Parses maximum major and minor`` (ver: Version) =
    let range = new IntegrationVersionRange()
    range.MaximumVersion <- $"{ver.Major}.{ver.Minor}"
    ver.Major = range.MaximumMajor &&
    ver.Minor = range.MaximumMinor

[<Property>]
let ``Parses maximum major and minor and patch`` (ver: Version) =
    let range = new IntegrationVersionRange()
    range.MaximumVersion <- $"{ver.Major}.{ver.Minor}.{ver.Patch}"
    ver.Major = range.MaximumMajor &&
    ver.Minor = range.MaximumMinor &&
    ver.Patch = range.MaximumPatch

File: tracer/test/Datadog.Trace.ClrProfiler.Managed.Tests/IntegrationVersionRangeTests.cs
// <copyright file="IntegrationVersionRangeTests.cs" company="Datadog">
// Unless explicitly stated otherwise all files in this repository are licensed under the Apache 2 License.
// This product includes software developed at Datadog (https://www.datadoghq.com/). Copyright 2017 Datadog, Inc.
// </copyright>

using System;
using NUnit.Framework;

namespace Datadog.Trace.ClrProfiler.Managed.Tests
{
    public class IntegrationVersionRangeTests
    {
        [Test]
        public void MinimumVersionTwoSetsResetsDefaultsForNonSpecifiedParts()
        {
            var range = new IntegrationVersionRange();
            range.MinimumVersion = "5.5.4";
            range.MinimumVersion = "6";
            Assert.AreEqual(expected: 6, actual: range.MinimumMajor);
            Assert.AreEqual(expected: 0, actual: range.MinimumMinor);
            Assert.AreEqual(expected: 0, actual: range.MinimumPatch);
        }

        [Test]
        public void ParsesMinimumMajor()
        {
            var range = new IntegrationVersionRange();
            range.MinimumVersion = "5";
            Assert.AreEqual(expected: 5, actual: range.MinimumMajor);
        }

        [Test]
        public void ParsesMinimumMajorAndMinor()
        {
            var range = new IntegrationVersionRange();
            range.MinimumVersion = "5.8";
            Assert.AreEqual(expected: 5, actual: range.MinimumMajor);
            Assert.AreEqual(expected: 8, actual: range.MinimumMinor);
        }

        [Test]
        public void ParsesMinimumMajorAndMinorAndPatch()
        {
            var range = new IntegrationVersionRange();
            range.MinimumVersion = "5.8.82";
            Assert.AreEqual(expected: 5, actual: range.MinimumMajor);
            Assert.AreEqual(expected: 8, actual: range.MinimumMinor);
            Assert.AreEqual(expected: 82, actual: range.MinimumPatch);
        }

        [Test]
        public void MaximumVersionTwoSetsResetsDefaultsForNonSpecifiedParts()
        {
            var range = new IntegrationVersionRange();
            range.MaximumVersion = "5.5.4";
            range.MaximumVersion = "6";
            Assert.AreEqual(expected: 6, actual: range.MaximumMajor);
            Assert.AreEqual(expected: ushort.MaxValue, actual: range.MaximumMinor);
            Assert.AreEqual(expected: ushort.MaxValue, actual: range.MaximumPatch);
        }

        [Test]
        public void ParsesMaximumMajor()
        {
            var range = new IntegrationVersionRange();
            range.MaximumVersion = "5";
            Assert.AreEqual(expected: 5, actual: range.MaximumMajor);
        }

        [Test]
        public void ParsesMaximumMajorAndMinor()
        {
            var range = new IntegrationVersionRange();
            range.MaximumVersion = "5.8";
            Assert.AreEqual(expected: 5, actual: range.MaximumMajor);
            Assert.AreEqual(expected: 8, actual: range.MaximumMinor);
        }

        [Test]
        public void ParsesMaximumMajorAndMinorAndPatch()
        {
            var range = new IntegrationVersionRange();
            range.MaximumVersion = "5.8.82";
            Assert.AreEqual(expected: 5, actual: range.MaximumMajor);
            Assert.AreEqual(expected: 8, actual: range.MaximumMinor);
            Assert.AreEqual(expected: 82, actual: range.MaximumPatch);
        }

        /// <summary>
        /// We want to be sure that any versions we specify are explicit, so we'll throw as soon as we know there is anything unclear.
        /// </summary>
        [Test]
        public void ThrowsExceptionForNonNumbers()
        {
            Exception exMin = null;
            Exception exMax = null;
            var range = new IntegrationVersionRange();

            try
            {
                range.MinimumVersion = "5.GARBAGE.82";
            }
            catch (Exception e)
            {
                exMin = e;
            }

            try
            {
                range.MaximumVersion = "5.35.MoreGarbage";
            }
            catch (Exception e)
            {
                exMax = e;
            }

            Assert.NotNull(exMin);
            Assert.NotNull(exMax);
        }
    }
}
